misc-scripts/xref_mapping/XrefParser/RefSeqGPFFParser.pm

```diff
@@ -17,340 +17,341 @@ limitations under the License.
-
-  $entrez_source_id = $self->get_source_id_for_source_name('EntrezGene', undef, $dbi);
-  $wiki_source_id = $self->get_source_id_for_source_name('WikiGene', undef, $dbi);
-
-  if($verbose){
-    print "RefSeq_peptide source ID = $peptide_source_id\n";
-    print "RefSeq_mRNA source ID = $mrna_source_id\n";
-    print "RefSeq_ncRNA source ID = $ncrna_source_id\n";
-    print "RefSeq_peptide_predicted source ID = $pred_peptide_source_id\n";
-    print "RefSeq_mRNA_predicted source ID = $pred_mrna_source_id\n" ;
-    print "RefSeq_ncRNA_predicted source ID = $pred_ncrna_source_id\n" ;
+
+  $self->{species_id} = $species_id;
+  $self->{dbi} = $dbi;
+  $self->{verbose} = $verbose;
+
+  # get RefSeq source ids
+  while (my ($source_prefix, $source_name) = each %{$REFSEQ_SOURCES}) {
+    $self->{source_ids}->{$source_name} = $self->get_source_id_for_source_name( $source_name, undef, $dbi )
+  }
+
+  # get extra source ids
+  $self->{source_ids}->{EntrezGene} = $self->get_source_id_for_source_name('EntrezGene', undef, $dbi);
+  $self->{source_ids}->{WikiGene} = $self->get_source_id_for_source_name('WikiGene', undef, $dbi);
+
+  # Retrieve existing RefSeq mRNA
+  $self->{refseq_ids} = { %{$self->get_valid_codes('RefSeq_mRNA', $species_id, $dbi)},
+      %{$self->get_valid_codes('RefSeq_mRNA_predicted', $species_id, $dbi)} };
+  $self->{entrez_ids} = $self->get_valid_codes('EntrezGene', $species_id, $dbi);
+  $self->{wiki_ids} = $self->get_valid_codes('WikiGene', $species_id, $dbi);
+
+  if ($verbose) {
+    for my $source_name (sort values %{$REFSEQ_SOURCES}) {
+      print "$source_name source ID = $self->{source_ids}->{$source_name}\n";
+    }
   }
 
-  (%entrez)     = %{$self->get_acc_to_label("EntrezGene",$species_id, undef, $dbi)};
+  # populate entrez gene id => label hash
+  $self->{entrez} = $self->get_acc_to_label('EntrezGene', $species_id, undef, $dbi);
+
+  # get the species name, prepare species related data checks
+  my %species2name = $self->species_id2name($dbi);
+  $species_name //= shift @{$species2name{$species_id}};
+
+  $self->{name2species_id} = { map{ $_=>$species_id } @{$species2name{$species_id}} };
+  my %species2tax  = $self->species_id2taxonomy($dbi);
+  $self->{taxonomy2species_id} = { map{ $_=>$species_id } @{$species2tax{$species_id}} };
+
+  # process the source files
+  foreach my $file (@{$files}) {
 
-    my @xrefs;
-    foreach my $file (@files) {
-        my $xrefs =
-          $self->create_xrefs( $file, $species_id, $verbose, $dbi, $species_name );
+    # type from the file (peptide or dna)
+    my $type = $self->type_from_file($file);
 
-        if ( !defined( $xrefs ) ) {
-            return 1;    #error
+    # get the file handler
+    my $refseq_fh = $self->get_filehandle($file);
+
+    if ( !defined $refseq_fh ) {
+      warn "WARNING: Can't open RefSeqGPFF file $file\n";
+      return;
+    }
+
+    # this will hold the array of xrefs to bulk insert
+    my $xrefs;
```

Comment:
`$xrefs` instead of `@xrefs` puzzled me at the beginning. You do you and it's of course equivalent, but I think "different things should look different" and I try to use @ and % wherever I can. At first I thought the code is e.g. passing `$xrefs` into a subroutine and pushing to it elsewhere - which would require using a ref - but happily it doesn't.

Also the comment about what the variable contains seems necessary, you could instead make the variable longer or make it say `@result` or something.

Reply:
a matter of style, I think @ and % are used far too often.
In this case @ would be equally good.

Reply:
I personally would explicitly assign [] to this in order to make this visually stand out as an arrayref, that said this _is_ a matter of style.

On the other hand, given this is used number-of-records-in-$refseq_fh times as an array and only once as a reference, it might make sense to make this an array after all.